File: Data/Java/trie insert and search.java
public class Trie {

    // Trie node
    static class TrieNode {
        TrieNode[] children = new TrieNode[26];  // For 'a' to 'z'
        boolean isEndOfWord;  // Marks end of a valid word
    }

    static class TrieImpl {
        TrieNode root = new TrieNode();

        // Insert word into trie
        public void insert(String word) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int index = c - 'a';

                // Create a new node if path doesn't exist
                if (node.children[index] == null)
                    node.children[index] = new TrieNode();

                node = node.children[index];  // Move to next node
            }
            node.isEndOfWord = true;  // Mark end of word
        }

        // Search if word exists in trie
        public boolean search(String word) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int index = c - 'a';

                if (node.children[index] == null)
                    return false;

                node = node.children[index];
            }
            return node.isEndOfWord;  // Return true if end is marked
        }
    }

    public static void main(String[] args) {
        TrieImpl trie = new TrieImpl();

        trie.insert("cat");
        trie.insert("car");
        trie.insert("cart");

        System.out.println("Search cat: " + trie.search("cat"));   // true
        System.out.println("Search cap: " + trie.search("cap"));   // false
    }
}

File: Data/Python/trie insert and search.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end_of_word = False  # True if this node ends a word

class Trie:
    """
    Trie (Prefix Tree):
    - Used for word insert and prefix/word search.
    - Time Complexity: O(L), where L = word length
    """
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        current = self.root
        for ch in word:
            if ch not in current.children:
                current.children[ch] = TrieNode()  # Create new node if not exists
            current = current.children[ch]
        current.end_of_word = True  # Mark end of word

    def search(self, word):
        current = self.root
        for ch in word:
            if ch not in current.children:
                return False  # Char not found
            current = current.children[ch]
        return current.end_of_word  # Return True only if it's a full word

# ✅ Example
trie = Trie()
trie.insert("cat")
trie.insert("car")
print("Search cat:", trie.search("cat"))  # True
print("Search cab:", trie.search("cab"))  # False
